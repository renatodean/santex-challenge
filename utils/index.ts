import { Item } from "../types/product";

/**
 * From a given product it will return the price as a formated string
 * @param product
 * @param quantity
 * @returns string
 */
export const getPrice = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return ((productPrice * (quantity || 1)) / 100).toFixed(2);
};

/**
 * From a given product it will return the price as number
 * @param product
 * @param quantity
 * @returns number
 */
export const getPriceNumber = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return (productPrice * (quantity || 1)) / 100;
};

/**
 * From a given query from router it will return the query value
 * @param query
 * @returns string
 */
export const getQueryValue = (query: any) => {
  if (Array.isArray(query)) {
    return query[0];
  } else {
    return query || "";
  }
};
