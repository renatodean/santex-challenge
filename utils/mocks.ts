export const mockedProducts = {
  products: {
    items: [
      {
        __typename: "Product",
        id: "1",
        name: "Laptop",
        slug: "laptop",
        description:
          "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
        featuredAsset: {
          __typename: "Asset",
          id: "1",
          name: "derick-david-409858-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 129900 },
          { __typename: "ProductVariant", price: 139900 },
          { __typename: "ProductVariant", price: 219900 },
          { __typename: "ProductVariant", price: 229900 },
        ],
      },
      {
        __typename: "Product",
        id: "2",
        name: "Tablet",
        slug: "tablet",
        description:
          'If the computer were invented today, what would it look like? It would be powerful enough for any task. So mobile you could take it everywhere. And so intuitive you could use it any way you wanted — with touch, a keyboard, or even a pencil. In other words, it wouldn’t really be a "computer." It would be Tablet.',
        featuredAsset: {
          __typename: "Asset",
          id: "2",
          name: "kelly-sikkema-685291-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/5a/kelly-sikkema-685291-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/b8/kelly-sikkema-685291-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 32900 },
          { __typename: "ProductVariant", price: 44500 },
        ],
      },
    ],
  },
};

export const mockedEmptyProducts = { products: { items: [] } };

export const mockedProduct = {
  product: {
    __typename: "Product",
    id: "1",
    name: "Laptop",
    slug: "laptop",
    description:
      "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
    featuredAsset: {
      __typename: "Asset",
      id: "1",
      name: "derick-david-409858-unsplash.jpg",
      type: "IMAGE",
      source:
        "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
      preview:
        "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
    },
    variants: [
      { __typename: "ProductVariant", price: 129900 },
      { __typename: "ProductVariant", price: 139900 },
      { __typename: "ProductVariant", price: 219900 },
      { __typename: "ProductVariant", price: 229900 },
    ],
  },
};

export const mockedStorageState = {
  showSnackbar: true,
  products: [
    {
      product: mockedProduct.product,
      quantity: 3,
    },
  ],
};
