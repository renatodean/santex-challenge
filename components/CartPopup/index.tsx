import React, { useMemo } from "react";
import styled from "@emotion/styled";
import { Item } from "../../types/product";
import EmptyCart from "../EmptyCart";
import CartProductDetail from "../CartProductDetail";
import { useRouter } from "next/router";
import { getPriceNumber } from "../../utils";

const Container = styled.div`
  display: flex;
  text-align: left;
  flex-direction: column;
  position: absolute;
  padding: 20px;
  width: 300px;
  height: auto;
  max-height: 50vh;
  background-color: white;
  border: 1px solid black;
  border-radius: 12px;
  z-index: 10;
  top: 50px;
  right: 10px;
  box-shadow: 0px 4px 2px rgba(0, 0, 0, 0.3);
`;

const CheckoutButton = styled.button`
  background-color: black;
  color: white;
  border: none;
  padding: 10px;
  border-radius: 8px;
  font-size: 20px;
  line-height: 24px;
  margin-top: 10px;
  margin-bottom: 10px;
  cursor: pointer;
  &:hover {
    background-color: rgba(1, 1, 1, 70%);
  }
`;

const CartPopupTitle = styled.h4`
  margin-top: 0;
  margin-bottom: 0;
`;

const ProductsContainer = styled.ul`
  padding-inline-start: 20px;
`;

const ProductsTotalContainer = styled.div`
  display: grid;
  grid-template-columns: 2fr 1fr;
  border-top: 0.5px solid rgba(0, 0, 0, 20%);
  padding-top: 10px;
  margin-bottom: 6px;
`;

const TotalInfo = styled.div`
  font-size: 20px;
  font-weight: 600;
`;

const TotalPrice = styled.div`
  font-size: 20px;
  font-weight: 600;
  text-align: right;
`;

interface CartPopupProps {
  products: {
    product: Item;
    quantity: number;
  }[];
}

const CartPopup = ({ products }: CartPopupProps) => {
  const router = useRouter();
  const totalPrice = useMemo(
    () =>
      products.reduce(
        (previous, product) =>
          previous + getPriceNumber(product.product, product.quantity),
        0
      ),
    [products]
  );

  return (
    <Container>
      {products && products.length ? (
        <>
          <CartPopupTitle>Products:</CartPopupTitle>
          <ProductsContainer>
            {products.map((product) => (
              <CartProductDetail key={product.product.slug} {...product} />
            ))}
          </ProductsContainer>
          <ProductsTotalContainer>
            <TotalInfo>Total:</TotalInfo>
            <TotalPrice>$ {totalPrice.toFixed(2)}</TotalPrice>
          </ProductsTotalContainer>
          <CheckoutButton
            tabIndex={0}
            onClick={() => router.push("/checkout")}
            onKeyPress={(event) => {
              if (event.key === "Enter") {
                router.push("/checkout");
              }
            }}
          >
            Checkout
          </CheckoutButton>
        </>
      ) : (
        <EmptyCart />
      )}
    </Container>
  );
};

export default CartPopup;
