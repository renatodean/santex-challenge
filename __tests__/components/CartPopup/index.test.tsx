import { render, screen, fireEvent } from "@testing-library/react";
import CartPopup from "../../../components/CartPopup";
import { mockedStorageState } from "../../../utils/mocks";

const routerPushSpy = jest.fn();

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: { q: "" },
      asPath: "",
      push: routerPushSpy,
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    };
  },
}));

describe("CartPopup", () => {
  it("renders the cartpopup with header", () => {
    render(<CartPopup products={mockedStorageState.products} />);

    expect(screen.getByText("Products:")).toBeInTheDocument();
  });

  it("should go to checkout when button it's clicked", () => {
    render(<CartPopup products={mockedStorageState.products} />);

    const checkoutButton = screen.getByTestId("checkout-button");
    fireEvent(
      checkoutButton,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(routerPushSpy).toHaveBeenCalled();
    expect(routerPushSpy).toHaveBeenCalledTimes(1);
    expect(routerPushSpy).toHaveBeenCalledWith("/checkout");
  });

  it("should go to checkout when button it's press with enter", () => {
    render(<CartPopup products={mockedStorageState.products} />);

    const checkoutButton = screen.getByTestId("checkout-button");
    fireEvent.keyPress(checkoutButton, { key: "Enter", code: "Enter", charCode: 13 });

    expect(routerPushSpy).toHaveBeenCalled();
    expect(routerPushSpy).toHaveBeenCalledTimes(2);
    expect(routerPushSpy).toHaveBeenCalledWith("/checkout");
  });
});

describe("CartPopup snapshot", () => {
  it("renders the cartpopup with products", () => {
    const { container } = render(
      <CartPopup products={mockedStorageState.products} />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the cartpopup without products", () => {
    const { container } = render(<CartPopup products={[]} />);
    expect(container).toMatchSnapshot();
  });
});
