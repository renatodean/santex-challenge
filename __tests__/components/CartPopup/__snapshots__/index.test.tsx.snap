// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPopup snapshot renders the cartpopup with products 1`] = `
<div>
  <div
    class="css-1q2mii1"
    data-testid="cart-popup"
  >
    <h4
      class="css-kcczxu"
    >
      Products:
    </h4>
    <ul
      class="css-yjavwb"
    >
      <li
        class="css-0"
      >
        <div
          class="css-27h8oo"
        >
          <div
            class="css-19dhafh"
          >
            3x 
            Laptop
          </div>
          <div
            class="css-6gq9lb"
          >
            $ 
            3897.00
          </div>
        </div>
      </li>
    </ul>
    <div
      class="css-14kbw0l"
    >
      <div
        class="css-hiimpy"
      >
        Total:
      </div>
      <div
        class="css-1hcwrky"
      >
        $ 
        3897.00
      </div>
    </div>
    <button
      class="css-1fwxpn2"
      data-testid="checkout-button"
      tabindex="0"
    >
      Checkout
    </button>
  </div>
</div>
`;

exports[`CartPopup snapshot renders the cartpopup without products 1`] = `
<div>
  <div
    class="css-1q2mii1"
    data-testid="cart-popup"
  >
    <div
      class="css-1kheigu"
    >
      <svg
        aria-hidden="true"
        class="StyledIconBase-ea9ulj-0 iLHuDN"
        fill="currentColor"
        focusable="false"
        height="36"
        viewBox="0 0 16 16"
        width="36"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.354 5.646a.5.5 0 1 0-.708.708L7.793 7.5 6.646 8.646a.5.5 0 1 0 .708.708L8.5 8.207l1.146 1.147a.5.5 0 0 0 .708-.708L9.207 7.5l1.147-1.146a.5.5 0 0 0-.708-.708L8.5 6.793 7.354 5.646z"
        />
        <path
          d="M.5 1a.5.5 0 0 0 0 1h1.11l.401 1.607 1.498 7.985A.5.5 0 0 0 4 12h1a2 2 0 1 0 0 4 2 2 0 0 0 0-4h7a2 2 0 1 0 0 4 2 2 0 0 0 0-4h1a.5.5 0 0 0 .491-.408l1.5-8A.5.5 0 0 0 14.5 3H2.89l-.405-1.621A.5.5 0 0 0 2 1H.5zm3.915 10L3.102 4h10.796l-1.313 7h-8.17zM6 14a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm7 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0z"
        />
      </svg>
      <h4
        class="css-1yltisf"
      >
        No products in the cart
      </h4>
      <span
        class="css-jfb38y"
      >
        Not sure where to start?
      </span>
      <span
        class="css-jfb38y"
      >
        Check out the homepage.
      </span>
    </div>
  </div>
</div>
`;
